Create P2PStateChannels.md

lukastanisic99 opened the pull request:
#### Project Abstract

Tech part of - 
https://docs.google.com/document/d/1DcDFScgeXijnXSGuaa30uOnYNtRm_9OnAsM97igrSwA/

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [x] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (Polkadot AssetHub (DOT, USDC & USDT) address in the application and bank details via email, if applicable).
- [x] I understand that an agreed upon percentage of each milestone will be paid in vested DOT, to the Polkadot address listed in the application.
- [x] I am aware that, in order to receive a grant, I (and the entity I represent) have to successfully complete a KYC/KYB check.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

keeganquigley commented:
Thanks for the application @lukastanisic99 very interesting tech indeed. My only question is, the milestone states, "Everything is done". Does this mean everything has already been implemented, or just that the milestone spreadsheet is finished?

lukastanisic99 commented:
@keeganquigley Well we did do a lot of the work already, but it's not complete and there's still a lot of work to be done and even later on maintaining and improving the system. The one milestone proposed was intended to be on completion of our V1 alpha and when everything is runnable. We could break it down further into smaller milestone that we can claim before the alpha if that's what you're referring.

semuelle reviewed (changes requested):
@lukastanisic99, we recently [updated our terms and template](https://github.com/w3f/Grants-Program/blame/master/applications/application-template.md#L127) to include a flexible DOT amount, can you please update your document to include this? Please keep in mind that this is a rather large amount you are asking for, so the bigger the DOT share, the better.

lukastanisic99 commented:
@semuelle We're fine with taking 100% in DOT with 50% being vested over 2 years, if all the DOT goes through a vesting process then we'll have to do 50% DOT and 50% USDT. Before I update the doc, can you clarify can we chose the % of DOT that has to be vested (minimum 50%) or all DOT received has to be vested, forcing us to take a portion in stablecoin? Normally we wouldn't mind taking 100% in vested DOT, but we have immediate costs..

semuelle commented:
Hi @lukastanisic99. Thanks for the quick reply. Currently, we only offer payment in fiat and stablecoins beside the vested (!) DOT portion. I will check internally if we can make an exception, but for now I would suggest we continue with this information.

lukastanisic99 commented:
Ok I've modified the doc to make the Milestone more clear (could break it down further if needed), added the USDT address on Asset Hub (same as DOT) and added the DOT % and indicated that we're willing to take 100% in DOT with 50% vested if that's possible. Maybe better to not squash future modifications so it's easier to keep track what has changed?

keeganquigley reviewed (changes requested):
Thanks @lukastanisic99 is there a chance that you'd be willing to come up with a proposal that is under $100k? As anything over this amount requires approval from the W3F council, and it could take a while to put it up for discussion. I will also say that we don't fund deliverables retroactively, so anything that has already been built could be removed from the milestone tables. Perhaps that would help to further reduce the scope/cost.

lukastanisic99 commented:
Hey @keeganquigley sure. I'll come up with a milestone that includes a minimal feature set so It's still runnable, but doesn't have all the features. I'll update the proposal later today or tomorrow. On the retroactive part - we did do some things (in private repos), we can always ctrl+A, delete and start from 0. Regardless that's not the point, there's still a lot of work to be done, so we'll try a figure out today a minimal set that works and update the proposal to reflect that, so we can continue and move on.

keeganquigley commented:
Thanks @lukastanisic99 sounds good; you can always break it up and apply for subsequent grants down the line.

lukastanisic99 commented:
I've updated the proposal to be within budget and explained what's included in the milestone. Take a look and let me know what are the next steps.

keeganquigley commented:
Sounds good @lukastanisic99 thanks for the changes. I will go ahead and mark the application as ready for review so other committee members can comment.

takahser commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+- **Payment Details:**
+  - **DOT**: 15ZzLCvcrZXDCddQnUGgzNUcMkSK4PugKdraQp8pFawoDWz1
+  - **USDT (Asset Hub)**: 15ZzLCvcrZXDCddQnUGgzNUcMkSK4PugKdraQp8pFawoDWz1
+- **[Level](https://github.com/w3f/Grants-Program/tree/master#level_slider-levels):** 2.5
```
```suggestion
- **[Level](https://github.com/w3f/Grants-Program/tree/master#level_slider-levels):** 3
```

takahser commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+- **Contact Name:** Luka Stanisic
+- **Contact Email:** lukastanisic2@gmail.com
+
+# Team's experience
+
+Luka - Background in engineering, protocol level of different DLTs (DAGs and blockchains), distributed systems at Microsoft Azure, Polkadot Blockchain Academy alumni with distinction.
+
+Dragan - Software engineer with 10+ years of experience.
+
+Aleksa Krstic - Senior UX/UI designer at another major DLT.
+
+Aleksa Gradinski - In crypto since 2017, CEO of Agroblock ($1M annual profit), helped in negotiations and expansion of leading web3 companies in Serbia.
+
+Gerard - Engineering degree holder with over a decade in high-end business development, overseeing multiple offices in the London property market. 15 years of professional poker playing and game theory expertise, bringing a strategic and analytical approach to business growth and a specific understanding of peer-to-peer commercial opportunities.
+
+Jeff - Applied cryptographer and mathematician at the Web3 Foundation. One of the lead protocol designers behind Polkadot and Kusama.
+
+Kasper - ex CFO Web3 Foundation and co-founder Polimec.
+
+Andrew - Researcher/Engineer at the Web3 Foundation
+
```
Could you please clarify which 7 people would be on your payroll?

> **Full-Time Equivalent (FTE):** 7 after aditional hires

takahser reviewed (changes requested):
@lukastanisic99 feel free to have a look at my comments. Apart from that, I'm currently talking to W3F-internal researchers, will be back soon with additional feedback.

lukastanisic99 commented:
@takahser I've merged your recommendations and clarified the section around FTE

burdges commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+The system is composed of a client side framework/SDK and on-chain smart contracts.
+
+Tech stack:
+On-chain: smart contracts (solidity)
```
The performance hit to polkadot here is minimal since they'll run rarely, given the whole system is a state channel.

burdges commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+
+Tech stack:
+On-chain: smart contracts (solidity)
+Off-chain (p2p) - written in typescript, holepunch for p2p and discovery, ethereumjs for running the EVM on client, ethers for manipulating and signing evm data and everything else is built from scratch (typescript). The entire design is state machine and runtime independent (can support substrate) - it works with native tooling. 
```
Is it worth clarifying that the application's solidiy run primarily on the user's side here, in the optimimistic case?

lukastanisic99 replied:
Yes it runs on the users side. Only during a dispute, when proving an incorrect state transition, the same bytecode gets invoked on-chain.

burdges commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+2) State machine replication - allowing connected peers to commit to state machines and allowing them to progress and replicate those state machines,
+3) Agreement tracking - tracking agreements on the order of transactions and state, managing forks, checking and collecting signatures, finalizing state
+4) Virtual (logical) clock - Having a local logical clock that’s synced to the DLT clock (timestamp) - allows peers to have a local perception of time that mimics the one of the DLT, but tolerates skewness to a degree. This is of course subjective and peers will never reach consensus on the clock nor with other peers nor with the DLT itself. This is very important and will be discussed further be ...
+5) Dispute handling - constructing fraud proofs when a peer deviates from the protocol,
```
I guess "fraud proof" is fine here, but mostly it just returns to running on-chain

The envisioned applciations have timeout information which after that concludes the execution can theoretically move back off-chain, although maybe that's an optimization not worth the development costs. I guess nothing much to change here.

lukastanisic99 replied:
Yes

burdges commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+	
+p2pTime -  the max amount of time a peer can take to decide, execute and broadcast their Tx to other peers,
+agreementTime - accounts for reasonable network delay, execution time and skewness of clocks. All agreements on the Tx (state transition) should happen within this time,
+chainFallbackTime - if p2p agreements fail within p2pTime + agreementTime, this is the last resort allowing peers to post on chain.
```
At least in the obvious applciations, you could move everything back off-chain after doing the on-chain stuff resumting from a timeout claim.  I'm not sure that optimization is worth the development costs though, exspecially if this rarely happens and everything eventually winds up compiled to polkavm.

burdges commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+
+Before explaining optimistic consensus, it’s important to understand the structure of a Transaction in this system:
+
+A transaction in this system references the exact state (previousStateHash) to which it is an input, and references the expected state post execution (stateHash). This resembles the logic behind UTXOs to a degree.
```
Are the required to be linearly ordered?  That's slightly more than what UTXOs do.  It's also what you expect when doing full state.  Assuming yes, anything else should proibably be future work.

burdges replied:
We'll punt on this for now, because the simplest thing alraedy makes sense for initial applications.

lukastanisic99 replied:
yes

burdges commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+
+A transaction in this system references the exact state (previousStateHash) to which it is an input, and references the expected state post execution (stateHash). This resembles the logic behind UTXOs to a degree.
+
+Under optimistic consensus we can’t force peers to collaborate, but it’s in their best interest to do so, since failure to collaborate at best invokes posting calldata on-chain and paying transaction fees and at worst results in a full slash when provable fraud can be detected. Collaboration includes signing (confirming) all valid transactions without delay (satisfying agreementTime) and producin ...
```
It's technically the applcations that must make it in their best interest to do so.  You cannot just port arbitrary smart contracts to a state channel, not without some incentives alterations.

burdges replied:
Anyways there is nothing to really change in the text here I guess.

lukastanisic99 replied:
The specific applications need to define the slashing/removal logic, but the general case of incentivizing cooperation comes from p2p being 'free' while on-chain execution requires a transaction fee in one way or the other.

burdges commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+Virtual Voting - Same concept as with blocks in a blockchain. Since blocks are cryptographically linked with past blocks through the parent hash, a block at block height K indirectly supports and thus votes for all previous K-1 blocks that precede it. In our case, a transaction issued is connected directly to the previous transaction through the previousStateHash and since this is recursive, it v ...
+
+Forks - 
+Forks ideally never happen and everyone cooperates fully p2p, without the need for the DLT ever to enforce a state, but in the case when the DLT has to step in, it creates a fork and sets the genesis state. The genesis state of a fork doesn’t require N signatures to be considered final since it’s enforced directly on the distributed ledger. Honest (non byzantine) peers always build on the latest  ...
```
The easiest solution here is for the UX to wait for BEEFY.  We do not need to do that right away thought.

burdges commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+3) Settlement - final settlement of any assets that entered into the state channel,
+4) Data Availability - Guaranteeing data availability for both state and transaction calldata if peer-to-peer agreements fail,
+5) Liveness guarantees - In the case that the state machine can’t progress peer-to-peer, the DLT can progress it and remove any participants who failed to progress it within timeout,
+6) Objective clock - provides a single global clock (timestamp) that’s uncontested and accepted as correct/true unconditionally. Peers sync their local logical clock to the DLT clock to mimic the DLT perception of time,
```
On 1+2+6.  Across all applications, the off-chain logic progresses by unanimous consensus, but individual steps have an author, and the others merely need to acknowledge them, while on-chain these acknowledgement steps can be replaced by a clock.  There is frequently an application specific part here too, in that if some turn logic dictates the next action's author, then if they abandon the channel then there must be some default action, even if only to abort for everyone.

lukastanisic99 replied:
yes

burdges commented on applications/P2PStateChannels.md:
```diff
@@ -0,0 +1,254 @@
+The minimal feature set (MFS) is the minimal product that's runnable and it includes:
+  - Permissionless p2p discovery and connection establishment through Holepunch,
+  - Opening the channel on-chain,
+  - Fully p2p state machine progression - running native smart contracts fully p2p.
```
We'd some question about this having only one milestone, no?  As I understand it, this milestone covers much of what's layed out in the Project Details section above.

I've left comments above on norner cases, some of which I've discossed with Luka, and on extensions, which mostly I did not previously discuss with Luka.  I think none of those comments represent architectural weaknesses, just stuff one might change before doing some applicaiton divferent form the envisioned applciations.

It's fairly involved, and Luka & Dragan have a fair amount of work here, with some of it like holepunch requiring a bunch of debugging work.  At the same time, they have largely worked out the architectural challenges in their prototyping.  Also, holepunch is a framework that "eventually works".  So they're not really in need of a bunch of smaller milestones.

takahser reviewed (changes requested):
@lukastanisic99 thanks for the updates and sorry for the delay. This ist starting to look better!

However, there are still a couple of things to fix.

i) Could you reply to @burdges' inline comments?
ii) Also, please transfer the milestones from the external sheet unto the md doc, so we can keep track of it.
iii) In addition to that, don't forget to add the [mandatory default deliverables](https://github.com/w3f/Grants-Program/blob/fe26185c4641cd25e0b1ede3a50ffe4aab5c9e11/applications/application-template.md#L135).

iv) I suggest to further break out the deliverables. For example, this sounds more like 3 deliverables to me:

    >"Blockchain Components:
    >Opening the Channel (MFS),
    >Liveness Gurantees (MFS),
    >Cryptographic primitives (MFS)"

v) Finally, I suggest to add more details to certain deliverables, such as the following:

    > Cryptography integration (MFS)


<img width="424" alt="image" src="https://github.com/user-attachments/assets/c76895dc-de43-47d2-b669-2118caab0113">

lukastanisic99 commented:
@takahser

i) Replied to comments,
ii) Will transfer the milestone from the sheet to the doc today,
iii) Will also add the mandatory deliverables today,
iv) Sure if you want we can break them down just for your accounting needs, but as discussed above with @burdges not worth the extra development cost + time focusing on mini deliverables and claiming one by one - we'll break it down as you proposed, but they'll be claimed all together.
v) sure I'll add more details to deliverable descriptions that might be unclear

lukastanisic99 commented:
@takahser done

takahser reviewed (changes requested):
@lukastanisic99 thx for the updates, but could you please add the deliverables into the table(s) rather than spreading them across different sections? I recommend to take a look at the [template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) for guidance.

lukastanisic99 commented:
@takahser done :D

takahser reviewed (changes requested):
@lukastanisic99 I'm happy with the scope and format now, but I'd like you to spec out the milestones in greater detail. Example: "All the logic required to open channels" is not specific enough. Similarly, "Cryptographic primitives to support threshold signature verification" is too generic - which algorithms are you going to use, are you going to re-use any existing work, etc. Also, for the "liveness" deliverable it would be good to have more details, such as techniques on how you'd plan to mitigate liveness issues.

Keep in mind that the deliverables act as a contract for us to evaluate whether a delivery satisfy the requirements. Hence, we'll need more details in the milestones.

lukastanisic99 commented:
@takahser I've made the requested modifications. The milestones just implement parts of the protocol as described in detail in the document and discussed with @burdges

takahser reviewed (dismissed):
@lukastanisic99 thx for the updates, happy to approve.

lukastanisic99 commented:
@keeganquigley I've sent you an email, but I guess I can also do it here - can we change the legal entity or can the KYC be on me, since the Swiss entity is not yet incorporated?

keeganquigley commented:
Hi @lukastanisic99 that's not an issue, but when there is no entity, all team members working on the grant need to complete KYC separately, from their own devices. If that's not an issue for your team, please use [this link instead](https://in.sumsub.com/idensic/l/#/uni_mjcB05gJZfh8dbr7) for KYC. 

Thanks!

keeganquigley commented:
@lukastanisic99 sure no problem, but in this case could you please update the team name on the application?

lukastanisic99 commented:
@keeganquigley updated the doc + all shareholders completed KYC

keeganquigley commented:
Great thanks @lukastanisic99 I see it is pending now. Will let you know when we get a result, it can take up to three days but is usually faster.

takahser reviewed (dismissed):
Re-approving.

keeganquigley commented:
Thanks @lukastanisic99 while waiting for KYB, could you also please add a [Legal Entity](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#legal-structure) section with the registered physical address as shown in the template? Thanks!

lukastanisic99 commented:
@keeganquigley done

takahser reviewed (approved):
Re-approving.

@lukastanisic99 please avoid force-pushing in case of future updates, to make it easier for us to review the changes. Note that the commits will be squashed anyway when merging to master.

lukastanisic99 commented:
@keeganquigley I've received an email that the KYB/KYC process in completed - double check on your end and let me know if there's anything else to do.

keeganquigley commented:
Thanks @lukastanisic99 I can confirm it passed, so now just waiting on the last two approvals. I pinged the committee again w/ the status update.

lukastanisic99 commented:
@semuelle and @PieWol can you take a look - we have 3/5 approvals

lukastanisic99 commented:
@laboon @SBalaguer @aeronbuchanan @rrtti @PieWol can anyone take a look and approve since we have 4/5 and want to finish this?

SBalaguer commented:
@lukastanisic99 I'm looking into it, should be able to provide feedback (if any) by end of day today.

SBalaguer commented:
@lukastanisic99 forgive me if I missed this from the long previous discussion, but where are you planning on deploying the smart contracts for the on-chain part?

lukastanisic99 commented:
@SBalaguer just to clarify, this is a public good SDK/framework to develop and deploy pure p2p state channels as described above. As such, we don't impose any restrictions on developers where they should deploy (nor could we if we wanted to - permissionless). The contracts can run in any EVM, so we assume most people would deploy their contracts to Moonbeam, Astar or any parachain that implements the EVM pallet (which many if not most already do).

We also designed it to be VM agnostic so it can later support the PVM, but too early to cover that now.

lukastanisic99 commented:
@SBalaguer just to clarify, this is a public good SDK/framework to develop and deploy pure p2p state channels as described above. As such, we don't impose any restrictions on developers where they should deploy (nor could we if we wanted to - permissionless). The contracts can run in any EVM, so we assume most people would deploy their contracts to Moonbeam, Astar or any parachain that implements the EVM pallet (which many if not most already do).

We also designed it to be VM agnostic so it can later support the PVM, but too early to cover that now.

lukastanisic99 commented:
Sorry for the double comment - seems github mobile glitched out

SBalaguer commented:
No worries at all @lukastanisic99, thanks for your answer. I'm looking forward to see how this develops! I strongly recommend ensuring compatibility with the PVM as stated on the comment.
